### ISSUE TITLE
Add a fail fast mode

### ISSUE THREAD

**Issue by mojavelinux** [asciidoctor/asciidoctor#803]
Add an option to enable fail fast, which will abort the processor at the first error-level message.

**Comment by elextr** [comment]
And an option to make warnings fail as well (from the Asciidoc list).

**Comment by mojavelinux** [comment]
Or, at the very least, the warnings should cause an exit mode of 1.

**Comment by mojavelinux** [comment]
> Or, at the very least, the warnings should cause an exit mode of 1.

We already have a way to do this now with the `--failure-level` CLI option.

**Comment by mojavelinux** [comment]
Open question 1: If multiple documents are passed to the `asciidoctor` command, should the command stop processing documents after the first document that hits the specified failure level? I think it should.

Open question 2: Do we really need to halt processing exactly where the first error is encountered, or is enough to just set the exit code? What would be the advantages / disadvantages of aborting processing early?

**Comment by elextr** [comment]
Q1

As a general rule, if multiple files are passed to an application and they are independent then failure of one should not stop processing of the others on the basis that it gives the user the most useful feedback as soon as possible, however if they are not independent then the first failure should stop the whole lot since the context may not be correct for the following ones.  

Its not clear from the docs what Asciidoctor does with multiple files, the [user manual](https://asciidoctor.org/docs/user-manual/#using-the-command-line-interface) doesn't allow multiple files and although the man page allows `FILE...` it does not discuss how the multiple files are treated, as one combined document or as separate documents.  That would determine which category above Asciidoctor falls into.

Q2
Only stop on the first unrecoverable error, if there is a reasonable assumption that the application can make that allows continued processing that can identify further errors, then that provides the user with the most useful feedback as soon as possible.  If an identified error is likely to cause subsequent false errors or just plain total confusion then its better to stop.  

Personal note, I _hate_ it when I get an error, fix it and then get another one in a few lines when running an application again, "why didn't you tell me that before!!!" :grin:

**Comment by mojavelinux** [comment]
> It's not clear from the docs what Asciidoctor does with multiple files

It's basically just shorthand for running the `asciidoctor` command multiple times. There's no inherent relationship between the files being processed.

**Comment by mojavelinux** [comment]
> Only stop on the first unrecoverable error, if there is a reasonable assumption that the application can make that allows continued processing that can identify further errors, then that provides the user with the most useful feedback as soon as possible.

I'd roughly describe that's how the log levels work in Asciidoctor today. An error means a significant problem in the content (like an unresolved include) but not a showstopper. A failure means processing cannot continue, like a missing converter (and processing is aborted).

Most problems in AsciiDoc are recoverable because the language is very forgiving. You might not get the output you expect, but nothing burned down ;)

**Comment by mojavelinux** [comment]
> Personal note, I hate it when I get an error, fix it and then get another one in a few lines when running an application again, "why didn't you tell me that before!!!" 

We share the same sentiment.

**Comment by elextr** [comment]
> Most problems in AsciiDoc are recoverable because the language is very forgiving.

Yes, mostly ignoring whatever caused the error (ie treat it as text) will allow the processor to continue, but for things that are document structure errors (like a section inside a block or an unmatched endif:: conditional directive) its probably indicative that there is a substantial error and the rest of the document structure is likely to parse wrongly.  Then stopping early might be best.  But its always a judgement call.

Also just because the parse completes, that doesn't say its possible to generate the output in the selected format, so thats clearly another fatal error.

> It's basically just shorthand for running the asciidoctor command multiple times.

So probably should run subsequent files, even if a preceding one failed.

> (that's where Antora comes in to the picture)

Looks neat, needs more advertising

**Comment by mojavelinux** [comment]
I definitely think it's time (or perhaps once we make the leap to 2.0.0) to start revisiting the severities. In an attempt to replicate AsciiDoc Python, we may have been too forgiving about certain errors. We should take each case on merit instead.

> So probably should run subsequent files, even if a preceding one failed.

:+1: 

> Looks neat, needs more advertising

I'd say we have just the right amount of users right now at this stage. Too many too soon and we end up getting flooded with people struggling with the same issues. You can be sure we'll be increasing the noise about it over course of the next year, esp after it has a chance to get lots of testing.

**Comment by mojavelinux** [comment]
Also covers #816.
